packages/babel-parser/src/parse-error.js

```diff
@@ -0,0 +1,211 @@
+// @flow
+
+import { Position } from "./util/location";
+import type { NodeBase } from "./types";
+import {
+  type ParseErrorCode,
+  ParseErrorCodes,
+  type ParseErrorCredentials,
+} from "./parse-error/credentials";
+
+const ArrayIsArray = Array.isArray;
+const {
+  assign: ObjectAssign,
+  defineProperty: ObjectDefineProperty,
+  getPrototypeOf: ObjectGetPrototypeOf,
+  keys: ObjectKeys,
+} = Object;
+
+type ToMessage<ErrorDetails> = (self: ErrorDetails) => string;
+
+const StandardMessage = Symbol("StandardMessage");
+
+// This should really be an abstract class, but that concept doesn't exist in
+// Flow, outside of just creating an interface, but then you can't specify that
+// it's a subclass of `SyntaxError`. You could do something like:
+//
+// interface IParseError<ErrorDetails> { ... }
+// type ParseError<ErrorDetails> = SyntaxError & IParseError<ErrorDetails>;
+//
+// But this is just a more complicated way of getting the same behavior, with
+// the added clumsiness of not being able to extends ParseError directly. So,
+// to keep things simple and prepare for a Typescript future, we just make it a
+// "private" superclass that we exclusively subclass:
+
+export class ParseError<ErrorDetails> extends SyntaxError {
+  code: ParseErrorCode;
+  reasonCode: string;
+
+  loc: Position;
+  details: ErrorDetails;
+
+  // There are no optional fields in classes in Flow, so we can't list these
+  // here: https://github.com/facebook/flow/issues/2859
+  // syntaxPlugin?: SyntaxPlugin
+  // missingPlugin?: string[]
+
+  constructor({
+    loc,
+    details,
+  }: {
+    loc: Position,
+    details: ErrorDetails,
+  }): ParseError<ErrorDetails> {
+    super();
+
+    this.loc = loc;
+
+    ObjectDefineProperty(this, "details", {
+      value: details,
+      enumerable: false,
+    });
+
+    // $FlowIgnore
+    if (details.missingPlugin) {
+      ObjectDefineProperty(this, "missingPlugin", {
+        get() {
+          return this.details.missingPlugin;
+        },
+        enumerable: true,
+      });
+    }
+
+    return this;
+  }
+
+  clone({
+    loc,
+    details,
+  }: {
+    loc?: Position,
+    details?: ErrorDetails,
+  } = {}) {
+    return new (ObjectGetPrototypeOf(this).constructor)({
```

Comment:
It might be, can't remember why I did that, I will check and change it to that if so.

Reply:
Ah, I remember now! I did that because `instance.constructor` *can* be mutated, but `Object.getPrototypeOf(instances).constructor` *can't* be mutated, so this guarantees we're getting the right class. If this is too defensive of a nit-pick, I can change it though.

Reply:
With the new system we don't do this at all anymore.